contracts/Core/StakeManager.sol

```diff
@@ -139,9 +139,9 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
         require(sAmount > 0, "Non-Positive Amount");
         IStakedToken sToken = IStakedToken(staker.tokenAddress);
         require(sToken.balanceOf(msg.sender) >= sAmount, "Invalid Amount");
+        emit Unstaked(epoch, stakerId, sAmount, staker.stake, block.timestamp);
```

Comment:
Hey 
Here either way there is no external call, so no reentrcy 
So why we are moving event up ?

Reply:
Yes, we dont have any external call but there is a statechange happening for lock

Reply:
Is it given that event should be emitted before state change, is there any pattern ?
I think it's just that event should be emitted before the external call, to avoid re-entracy 
```
Reentrancy in BlockManager.claimBlockReward() (contracts/Core/BlockManager.sol#110-125):
	External calls:
	- rewardManager.giveBlockReward(stakerId,epoch) (contracts/Core/BlockManager.sol#123)
	Event emitted after the call(s):
	- BlockConfirmed(epoch,proposerId,proposedBlocks[epoch][blockId].medians,block.timestamp) (contracts/Core/BlockManager.sol#124)
```
If we start emitting events before state change, what should we emit in event, old data :)

Reply:
For example 
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cec4f2ef57495d8b1742d62846da212515d99dd5/contracts/token/ERC20/ERC20.sol#L237

Reply:
In our case yes, because we are emitting the resultant after state change.

Reply:
actually here we do have external call, parameters.withdrawLockPeriod())